mlir/lib/Target/LLVMIR/ModuleTranslation.cpp

```diff
@@ -1042,17 +1046,77 @@ LogicalResult ModuleTranslation::convertGlobals() {
   for (auto op : getModuleBody(mlirModule).getOps<LLVM::GlobalOp>()) {
     if (Block *initializer = op.getInitializerBlock()) {
       llvm::IRBuilder<> builder(llvmModule->getContext());
+
+      int numConstantsHit = 0;
+      int numConstantsErased = 0;
+      DenseMap<llvm::ConstantAggregate *, int> constantAggregateUseMap;
+
       for (auto &op : initializer->without_terminator()) {
         if (failed(convertOperation(op, builder)) ||
             !isa<llvm::Constant>(lookupValue(op.getResult(0))))
           return emitError(op.getLoc(), "unemittable constant value");
+
+        // When emitting an LLVM constant, a new constant is created and the old
+        // constant may become dangling and take space. We should remove the
+        // dangling constants to avoid memory explosion especially for constant
+        // arrays whose number of elements is large.
+        // Because multiple operations may refer to the same constant, we need
+        // to count the number of uses of each constant array and remove it only
+        // when the count becomes zero.
+        if (op.getNumResults() == 1) {
+          Value result = op.getResult(0);
+          auto cst = dyn_cast<llvm::ConstantAggregate>(lookupValue(result));
+          if (!cst)
+            continue;
+          numConstantsHit++;
+          auto iter = constantAggregateUseMap.find(cst);
+          int numUsers = std::distance(result.use_begin(), result.use_end());
+          if (iter == constantAggregateUseMap.end())
+            constantAggregateUseMap.try_emplace(cst, numUsers);
+          else
+            iter->second += numUsers;
+        }
+        for (Value v : op.getOperands()) {
+          auto cst = dyn_cast<llvm::ConstantAggregate>(lookupValue(v));
+          if (!cst)
+            continue;
+          auto iter = constantAggregateUseMap.find(cst);
+          assert(iter != constantAggregateUseMap.end() && "constant not found");
+          iter->second--;
+          if (iter->second == 0) {
+            cst->removeDeadConstantUsers();
+            if (cst->user_empty()) {
+              cst->destroyConstant();
+              numConstantsErased++;
+            }
+            constantAggregateUseMap.erase(iter);
+          }
+        }
       }
```

Comment:
I feel I'm missing something, why can't we have a `DenseSet<llvm::ConstantAggregate *> constantAggregates;` that we keep around and every iteration of the outer loop we could just do:
```
for (llvm::ConstantAggregate *cst : constantAggregates) {
  cst->removeDeadConstantUsers();
  if (cst->user_empty()) {
    cst->destroyConstant(); // TODO: add removing from the DenseSet
  }
}
```

Reply:
Actually this is done also line 1104, but I don't get why we need the dance around `constantAggregateUseMap` here?

Reply:
> I feel I'm missing something, why can't we have a `DenseSet<llvm::ConstantAggregate *> constantAggregates;` that we keep around and every iteration of the outer loop we could just do:
> 
> ```
> for (llvm::ConstantAggregate *cst : constantAggregates) {
>   cst->removeDeadConstantUsers();
>   if (cst->user_empty()) {
>     cst->destroyConstant(); // TODO: add removing from the DenseSet
>   }
> }
> ```

My understanding is that we're converting all the ops in intializer to llvm::Constant.
When a later op is converted into ConstantAggregate, it will add use of the constant converted from early op.
So we cannot remove a Constant unless all the ops which use it are converted.
Only check user_empty is not enough because the user might not yet be generated.

Reply:
Ah I get it: and how critical is it to garbage collect during the interpretation of the initializer block instead of just at the end of the block?
(line 1104)

Reply:
The worst block is building a constant with type [75525 x [100 x f32]] for my test case.
I can test only do the removing at the end of the block, code will be much simplified if it works.

Reply:
Also: isn't it something that MLIR can constant fold?

Reply:
The constant will fold, but the input constant is still in memory.
And too many temp constants are created which eats all the memory.

I tried to avoid creating these insertvalue which will be converted to temp constant.

Here's what I got with experiment

```

  llvm.mlir.global private @krnl_global_28(dense<["", "ab", "xyz", ...]> : tensor<75525x!krnl.string>) {addr_space = 0 : i32, alignment = 8 : i64} : !llvm.array<75525 x i64>

```
It crashes because !llvm.array<75525 x i64> is not correct type.

Any suggestion for how to express array of string for llvm dialect?

Reply:
Wouldn't it be just a concat of all the strings into a single dense attribute of the right type?

Reply:
> The worst block is building a constant with type [75525 x [100 x f32]] for my test case. I can test only do the removing at the end of the block, code will be much simplified if it works.

Removing at the end not work :( One block already used all of the memory.

Reply:
> Wouldn't it be just a concat of all the strings into a single dense attribute of the right type?

Thanks for this suggestion. It saves a lot of time for create globalOp for each string!

But to create the array of string, I still have to create a lot of insertvalue.

Is it possible to express the following llvm ir with LLVM dialect?
```

@c = dso_local global [7 x i8] c"abcxby\00", align 1
@a = dso_local global [3 x ptr] [ptr @c, ptr getelementptr (i8, ptr @c, i64 2), ptr getelementptr (i8, ptr @c, i64 5)]
```
Then there will be no insertvalue needed.